File: examplepatches/ch3thanhs/stylus/extensions/extension/src/main/java/app/stylus/extension/bitwarden/patches/ForceSystemFontPatch.java
package app.stylus.extension.bitwarden.patches;

import android.graphics.Typeface;

@SuppressWarnings("unused")
public final class ForceSystemFontPatch {

    private ForceSystemFontPatch() {
    }

    /**
     * Injection point.
     *
     * <p>Returns the system default {@link Typeface} for the given style so the
     * app renders using the device's system font instead of bundled DM Sans/Roboto files.
     *
     * @param style Typeface style (Typeface.NORMAL, BOLD, ITALIC, or BOLD_ITALIC).
     * @return The system default typeface for {@code style}.
     */
    public static Typeface getSystemTypeface(int style) {
        return Typeface.create(Typeface.DEFAULT, style);
    }
}

File: examplepatches/ch3thanhs/stylus/patches/src/main/kotlin/app/stylus/patches/bitwarden/font/ForceSystemFontPatch.kt
package app.stylus.patches.bitwarden.font

import app.morphe.patcher.extensions.InstructionExtensions.addInstructionsWithLabels
import app.morphe.patcher.patch.bytecodePatch
import app.stylus.patches.bitwarden.shared.Constants.COMPATIBILITY_BITWARDEN

private const val EXTENSION_CLASS =
    "Lapp/stylus/extension/bitwarden/patches/ForceSystemFontPatch;"

@Suppress("unused")
val forceSystemFontPatch = bytecodePatch(
    name = "Force system font (Bitwarden)",
    description = "Renders the app using the device's system font instead of DM Sans / Roboto.",
    default = true,
) {
    compatibleWith(COMPATIBILITY_BITWARDEN)

    extendWith("extensions/extension.mpe")

    execute {
        val targetMethod = resolveTypefaceCompatCreateFromResourcesFontFileTarget()

        // Param mapping changes between legacy and modern method shapes.
        targetMethod.method.addInstructionsWithLabels(
            0,
            when (targetMethod.variant) {
                TypefaceCompatCreateFromResourcesFontFileVariant.LEGACY_STATIC ->
                    """
                        invoke-static { p4 }, $EXTENSION_CLASS->getSystemTypeface(I)Landroid/graphics/Typeface;
                        move-result-object v0
                        if-eqz v0, :original
                        return-object v0
                        :original
                        nop
                    """

                TypefaceCompatCreateFromResourcesFontFileVariant.MODERN_VIRTUAL ->
                    """
                        invoke-static { p5 }, $EXTENSION_CLASS->getSystemTypeface(I)Landroid/graphics/Typeface;
                        move-result-object v0
                        if-eqz v0, :original
                        return-object v0
                        :original
                        nop
                    """
            }
        )
    }
}